File: SimpleTextEditor/SimpleTextEditor.js
function processData(input) {
    //Enter your code here
     const lines = input.trim().split('\n');
    const q = parseInt(lines[0]);
    
    let current = '';
    const history = [];

    for (let i = 1; i <= q; i++) {
        const [op, arg] = lines[i].split(' ');

        switch (op) {
            case '1': // Append
                history.push(current);
                current += arg;
                break;

            case '2': // Delete
                history.push(current);
                const k = parseInt(arg);
                current = current.slice(0, -k);
                break;

            case '3': // Print
                const index = parseInt(arg) - 1;
                console.log(current[index]);
                break;

            case '4': // Undo
                if (history.length > 0) {
                    current = history.pop();
                }
                break;
        }
    }
} 

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
   processData(_input);
});

File: SimpleTextEditor/SimpleTextEditor.java
import java.io.*;
import java.util.*;

public class Solution {

    public static void main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        Scanner sc = new Scanner(System.in);
        int q = sc.nextInt();
        sc.nextLine();
        StringBuilder current = new StringBuilder();
        Stack<String> history = new Stack<>();
        
        for(int i=0; i<q; i++){
            String[] parts = sc.nextLine().split(" ");
            String cmd = parts[0]; // first element is the operation
            
            switch(cmd){
                case "1":
                    history.push(current.toString());
                    current.append(parts[1]);
                    break;
                 case "2":
                    history.push(current.toString());
                    int k = Integer.parseInt(parts[1]);
                    current.setLength(current.length() - k); //set to the first current.length*() -k
                    break;
                 case "3":
                    int input = Integer.parseInt(parts[1])-1;
                    System.out.println(current.charAt(input));
                    break;
                 case "4":
                    if(!history.isEmpty()){
                        current = new StringBuilder(history.pop());
                    }
                    break;
                
            }
        }
        sc.close();
    }
}
